File: study/lua/src/app/data/EquipmentUnitData.lua
--
-- Author: Liangxu
-- Date: 2017-04-06 14:17:45
-- 装备单元数据
local BaseData = require("app.data.BaseData")
local EquipmentUnitData = class("EquipmentUnitData", BaseData)

local schema = {}
schema["id"] = {"number", 0} --唯一Id
schema["base_id"] = {"number", 0} --静态表Id
schema["level"] = {"number", 0} --强化等级
schema["star"] = {"number", 0} --星级
schema["rank"] = {"number", 0} --品质
schema["money"] = {"number", 0} --强化消耗的钱
schema["time"] = {"number", 0} --获得时间
schema["user_id"] = {"number", 0} --玩家id
schema["r_level"] = {"number", 0} --精炼等级
schema["r_exp"] = {"number", 0} --精炼经验
schema["all_exp"] = {"number", 0} --精炼消耗的经验
schema["config"] = {"table", {}} --配置表信息
schema["yokeRelation"] = {"boolean", false} --是否有羁绊关系
schema["materials"] = {"table", {}} -- 界限突破消耗材料
schema["recycle_materials"] = {"table", {}} -- 界限突破回收材料、
schema["jades"] = {"table", {}} -- 装备玉石、
schema["change"] = {"number", 0} --装备是否有重新强化或精炼过
schema["last_jades"] = {"table", {}}
EquipmentUnitData.schema = schema

function EquipmentUnitData:ctor(properties)
    EquipmentUnitData.super.ctor(self, properties)
    self._isUser = true
end

function EquipmentUnitData:clear()
end

function EquipmentUnitData:reset()
end

function EquipmentUnitData:updateData(data)
    self:backupJades()
    self:setProperties(data)
    local config = require("app.config.equipment").get(data.base_id)
    assert(config, "equipmentConfig can't find base_id = " .. tostring(data.base_id))
    self:setConfig(config)
end

function EquipmentUnitData:getPos()
    local id = self:getId()
    local data = G_UserData:getBattleResource():getEquipDataWithId(id)
    if data then
        return data:getPos()
    else
        return nil
    end
end

function EquipmentUnitData:getSlot()
    local id = self:getId()
    local data = G_UserData:getBattleResource():getEquipDataWithId(id)
    if data then
        return data:getSlot()
    else
        return nil
    end
end

function EquipmentUnitData:isInBattle()
    local id = self:getId()
    local data = G_UserData:getBattleResource():getEquipDataWithId(id)
    if data == nil then
        return false
    else
        return true
    end
end

-- 是否镶嵌玉石
function EquipmentUnitData:isInjectJadeStone()
    local jades = self:getJades()
    for i = 1, #jades do
        if jades[i] > 0 then
            return true
        end
    end
    return false
end

--是否强化过
function EquipmentUnitData:isDidStrengthen()
    return self:getLevel() > 1
end

--是否精炼过
function EquipmentUnitData:isDidRefine()
    return self:getR_level() > 0
end

--是否养成过
function EquipmentUnitData:isDidTrain()
    local isDidStrengthen = self:isDidStrengthen()
    local isDidRefine = self:isDidRefine()
    if isDidStrengthen or isDidRefine then
        return true
    else
        return false
    end
end

-- 获取相应槽位获取玉石系统id
function EquipmentUnitData:getJadeSysIds()
    local jadeSysIds = {}
    local jades = self:getJades()
    for i = 1, #jades do
        if jades[i] > 0 then
            local unitData = G_UserData:getJade():getJadeDataById(jades[i])
            jadeSysIds[i] = unitData:getConfig().id
        else
            jadeSysIds[i] = 0
        end
    end
    return jadeSysIds
end

--------------------------------------------------
---TODO：因该是通过ID==0来判断
function EquipmentUnitData:setIsUserEquip(isUser)
    self._isUser = isUser
end

function EquipmentUnitData:isUserEquip()
    return self._isUser
end

-- 是否界限突破过
function EquipmentUnitData:isLimitUp()
    local bres = false
    for k, v in pairs(self:getMaterials()) do
        if v > 0 then
            bres = true
        end
    end
    if #self:getRecycle_materials() > 0 then
        bres = true
    end
    return bres
end

-- 是否为白板装备
function EquipmentUnitData:isBlackPlat()
    local flag = self:getMoney() == 0
    flag = flag and self:getAll_exp() == 0
    flag = flag and #self:getRecycle_materials() == 0
    flag = flag and not self:isDidTrain()
    for k, v in pairs(self:getMaterials()) do
        flag = flag and v == 0
    end
    for k, v in pairs(self:getJades()) do
        flag = flag and v == 0
    end
    return flag
end

-- 判断是否已经镶嵌了两颗同玉石
function EquipmentUnitData:isHaveTwoSameJade(jadeId)
    local jadeUnit = G_UserData:getJade():getJadeDataById(jadeId)
    local slots = {}
    if not jadeUnit then
        return false
    end
    local jades = self:getJades()
    local count = 0
    for i = 1, #jades do
        if jades[i] > 0 then
            local unitData = G_UserData:getJade():getJadeDataById(jades[i])
            if jadeUnit:getConfig().id == unitData:getConfig().id then
                count = count + 1
                slots[i] = true
            end
        end
    end
    return count >= 2, slots
end

function EquipmentUnitData:backupJades()
    local lastJade = {}
    local jades = self:getJades()
    for i = 1, #jades do
        table.insert(lastJade, jades[i])
    end
    self:setLast_jades(lastJade)
end

-- 属性玉石是否镶嵌满
function EquipmentUnitData:isFullAttrJade()
    local config = self:getConfig()
    if config.suit_id == 0 then
        return false
    end
    local slotinfo = string.split(config.inlay_type, "|")
    local jades = self:getJades()
    for i = 2, #slotinfo do
        if tonumber(slotinfo[i]) > 0 then
            if jades[i] == 0 then
                return false
            else
                if not self:isActiveJade(jades[i]) then
                    return false
                end
            end
        end
    end
    return true
end

-- 玉石是否生效
function EquipmentUnitData:isActiveJade(id)
    local unitData = G_UserData:getJade():getJadeDataById(id)
    local UserDataHelper = require("app.utils.UserDataHelper")
    local _, heroBaseId = UserDataHelper.getHeroBaseIdWithEquipId(self:getId())
    local isSuitable = unitData:isSuitableHero(heroBaseId)
    return isSuitable
end

-- 所有玉石是否镶嵌满
function EquipmentUnitData:isFullJade()
    local config = self:getConfig()
    local slotinfo = string.split(config.inlay_type, "|")
    if tonumber(slotinfo[1]) == 0 then
        return false
    end
    local jades = self:getJades()
    for i = 1, #slotinfo do
        if tonumber(slotinfo[i]) > 0 then
            if jades[i] == 0 then
                return false
            else
                if not self:isActiveJade(jades[i]) then
                    return false
                end
            end
        end
    end
    return true
end

-- 设置查看用户数据时的装备玉石sys_id
function EquipmentUnitData:setUserDetailJades(slot, sysId)
    self._userDetailJades = self._userDetailJades or {}
    self._userDetailJades[slot] = sysId
end

function EquipmentUnitData:getUserDetailJades()
    return self._userDetailJades 
end

function EquipmentUnitData:getJadeSlotNums()
    local config = self:getConfig()
    if config.suit_id == 0 then
        return 0
    end
    local count = 0
    local slotinfo = string.split(config.inlay_type, "|")
    for _, value in pairs(slotinfo) do
        if tonumber(value) > 0 then
            count = count + 1
        end
    end
    return count
end
-------------------------------------------

return EquipmentUnitData

File: study/ts/script/data/EquipmentUnitData.ts

import { ConfigNameConst } from "../const/ConfigNameConst";
import { BaseData } from "./BaseData";
import { G_ConfigLoader, G_UserData } from "../init";
import { JadeStoneData } from "./JadeStoneData";
import { UserDataHelper } from "../utils/data/UserDataHelper";

var schema = {};
schema['id'] = [
    'number',
    0
];
schema['base_id'] = [
    'number',
    0
];
schema['level'] = [
    'number',
    0
]
schema['star'] = [
    'number',
    0
]
schema['rank'] = [
    'number',
    0
]
schema['money'] = [
    'number',
    0
]
schema['time'] = [
    'number',
    0
]
schema['user_id'] = [
    'number',
    0
]
schema['r_level'] = [
    'number',
    0
]
schema['r_exp'] = [
    'number',
    0
]
schema['all_exp'] = [
    'number',
    0
]
schema['config'] = [
    'object',
    {}
]
schema['yokeRelation'] = [
    'boolean',
    false
]
schema['materials'] = [
    'object',
    []
]
schema['recycle_materials'] = [
    'object',
    []
]
schema['jades'] = [
    'object',
    []
]
schema['change'] = [
    'number',
    0
]
schema['last_jades'] = [];

export interface EquipmentUnitData {
    getId(): number
    setId(value: number): void
    getLastId(): number
    getBase_id(): number
    setBase_id(value: number): void
    getLastBase_id(): number
    getLevel(): number
    setLevel(value: number): void
    getLastLevel(): number
    getStar(): number
    setStar(value: number): void
    getLastStar(): number
    getRank(): number
    setRank(value: number): void
    getLastRank(): number
    getMoney(): number
    setMoney(value: number): void
    getLastMoney(): number
    getTime(): number
    setTime(value: number): void
    getLastTime(): number
    getUser_id(): number
    setUser_id(value: number): void
    getLastUser_id(): number
    getR_level(): number
    setR_level(value: number): void
    getLastR_level(): number
    getR_exp(): number
    setR_exp(value: number): void
    getLastR_exp(): number
    getAll_exp(): number
    setAll_exp(value: number): void
    getLastAll_exp(): number
    getConfig(): any
    setConfig(value: any): void
    getLastConfig(): any
    isYokeRelation(): boolean
    setYokeRelation(value: boolean): void
    isLastYokeRelation(): boolean
    getMaterials(): number[]
    setMaterials(value: number[]): void
    getLastMaterials(): number[]
    getRecycle_materials(): number[]
    setRecycle_materials(value: number[]): void
    getLastRecycle_materials(): number[]
    getJades(): number[]
    setJades(value: number[]): void
    getLastJades(): number[]
    getChange(): number
    setChange(value: number): void
    getLastChange(): number
    getLast_jades(): number[]
    setLast_jades(value: number[]): void
    getLastLast_jades(): number[]
}
export class EquipmentUnitData extends BaseData {
    public static schema = schema;
    private _isUser = true;
    private _userDetailJades;
    updateData(data) {
        this.backupJades();
        this.setProperties(data);
        var config = G_ConfigLoader.getConfig(ConfigNameConst.EQUIPMENT).get(data.base_id);
        console.assert(config, 'equipmentConfig can\'t find base_id = ' + (data.base_id));
        this.setConfig(config);
    }
    getPos() {
        var id = this.getId();
        var data = G_UserData.getBattleResource().getEquipDataWithId(id);
        if (data) {
            return data.getPos();
        } else {
            return null;
        }
    }
    getSlot() {
        var id = this.getId();
        var data = G_UserData.getBattleResource().getEquipDataWithId(id);
        if (data) {
            return data.getSlot();
        } else {
            return null;
        }
    }
    isInBattle() {
        var id = this.getId();
        var data = G_UserData.getBattleResource().getEquipDataWithId(id);
        if (data == null) {
            return false;
        } else {
            return true;
        }
    }
    isInjectJadeStone() {
        var jades = this.getJades();
        for (let i = 0; i < jades.length; i++) {
            if (jades[i] > 0) {
                return true;
            }
        }
        return false;
    }
    isDidStrengthen() {
        return this.getLevel() > 1;
    }
    isDidRefine() {
        return this.getR_level() > 0;
    }
    isDidTrain() {
        var isDidStrengthen = this.isDidStrengthen();
        var isDidRefine = this.isDidRefine();
        if (isDidStrengthen || isDidRefine) {
            return true;
        } else {
            return false;
        }
    }
    getJadeSysIds() {
        var jadeSysIds = [];
        var jades = this.getJades();
        for (var i = 0; i < jades.length; i++) {
            if (jades[i] > 0) {
                var unitData = G_UserData.getJade().getJadeDataById(jades[i]);
                jadeSysIds.push(unitData.getConfig().id);
            } else {
                jadeSysIds.push(0);
            }
        }
        return jadeSysIds;
    }
    setIsUserEquip(isUser) {
        this._isUser = isUser;
    }
    isUserEquip() {
        return this._isUser;
    }
    isLimitUp() {
        var bres = false;
        for (var k in this.getMaterials()) {
            var v = this.getMaterials()[k];
            if (v > 0) {
                bres = true;
            }
        }
        if (this.getRecycle_materials().length > 0) {
            bres = true;
        }
        return bres;
    }
    isBlackPlat() {
        var flag = this.getMoney() == 0;
        flag = flag && this.getAll_exp() == 0;
        flag = flag && this.getRecycle_materials().length == 0;
        flag = flag && !this.isDidTrain();
        for (var k in this.getMaterials()) {
            var v = this.getMaterials()[k];
            flag = flag && v == 0;
        }
        for (var k in this.getJades()) {
            var v = this.getJades()[k];
            flag = flag && v == 0;
        }
        return flag;
    }
    isHaveTwoSameJade(jadeId) {
        var jadeUnit = G_UserData.getJade().getJadeDataById(jadeId);
        var slots = {};
        if (!jadeUnit) {
            return [false];
        }
        var jades = this.getJades();
        var count = 0;
        for (var i = 0; i < jades.length; i++) {
            if (jades[i] > 0) {
                var unitData = G_UserData.getJade().getJadeDataById(jades[i]);
                if (jadeUnit.getConfig().id == unitData.getConfig().id) {
                    count = count + 1;
                    slots[i] = true;
                }
            }
        }
        return [
            count >= 2,
            slots
        ];
    }
    backupJades() {
        var lastJade = [];
        var jades = this.getJades();
        for (var i = 0; i < jades.length; i++) {
            lastJade[i] = jades;
        }
        this.setLast_jades(lastJade);
    }
    isFullAttrJade() {
        var config = this.getConfig();
        if (config.suit_id == 0) {
            return false;
        }
        var slotinfo = config.inlay_type.split('|');
        var jades = this.getJades();
        for (var i = 0; i < slotinfo.length; i++) {
            if (parseFloat(slotinfo[i]) > 0) {
                if (jades[i] == 0) {
                    return false;
                } else {
                    if (!this.isActiveJade(jades[i])) {
                        return false;
                    }
                }
            }
        }
        return true;
    }
    isActiveJade(id) {
        var unitData = G_UserData.getJade().getJadeDataById(id);
        var [_, heroBaseId] = UserDataHelper.getHeroBaseIdWithEquipId(this.getId());
        var isSuitable = unitData.isSuitableHero(heroBaseId);
        return isSuitable;
    }
    isFullJade() {
        var config = this.getConfig();
        var slotinfo = config.inlay_type.split('|');
        if (parseFloat(slotinfo[1]) == 0) {
            return false;
        }
        var jades = this.getJades();
        for (var i = 0; i < slotinfo.length; i++) {
            if (parseFloat(slotinfo[i]) > 0) {
                if (jades[i] == 0) {
                    return false;
                } else {
                    if (!this.isActiveJade(jades[i])) {
                        return false;
                    }
                }
            }
        }
        return true;
    }
    setUserDetailJades(slot, sysId) {
        this._userDetailJades = this._userDetailJades || {};
        this._userDetailJades[slot] = sysId;
    }
    getUserDetailJades() {
        return this._userDetailJades;
    }
    getJadeSlotNums() {
        var config = this.getConfig();
        if (config.suit_id == 0) {
            return 0;
        }
        var count = 0;
        var slotinfo = config.inlay_type.split('|');
        for (var i in slotinfo) {
            var value = slotinfo[i];
            if (parseFloat(value) > 0) {
                count = count + 1;
            }
        }
        return count;
    }
}